File: TableturfBattleServer/Stage.cs
using Newtonsoft.Json;

namespace TableturfBattleServer;
public class Stage(string name, Space[,] grid, Point[][] startSpaces) {
	public string Name { get; } = name ?? throw new ArgumentNullException(nameof(name));
	[JsonProperty]
	internal Space[,] Grid = grid ?? throw new ArgumentNullException(nameof(grid));
	/// <summary>
	///		The lists of starting spaces on this stage.
	/// </summary>
	/// <remarks>
	///		The smallest list with at least as many spaces as players in the game will be used.
	///		For example, if there is a list of 3 and a list of 4, the list of 3 will be used for 2 or 3 players, and the list of 4 will be used for 4 players.
	/// </remarks>
	[JsonProperty]
	internal Point[][] StartSpaces = startSpaces ?? throw new ArgumentNullException(nameof(startSpaces));

	[JsonIgnore]
	public int MaxPlayers => this.StartSpaces.Max(a => a.Length);
}

File: TableturfBattleClient/src/Stage.ts
class Stage {
	name: string;
	grid: readonly (readonly Space[])[];
	startSpaces: Point[][];

	constructor(name: string, grid: Space[][], startSpaces: Point[][]) {
		this.name = name;
		this.grid = grid;
		this.startSpaces = startSpaces;
	}

	static fromJson(obj: any) {
		return new Stage(obj.name, obj.grid, obj.startSpaces);
	}

	get maxPlayers() { return Math.max(...this.startSpaces.map(a => a.length)); }

	getStartSpaces(numPlayers: number) {
		let list = null as Point[] | null;
		for (const list2 of this.startSpaces) {
			if (list2.length >= numPlayers && (list == null || list2.length < list.length))
				list = list2;
		}
		return list!;
	}

	copyGrid() {
		return Array.from(this.grid, a => a.slice(0));
	}
}
